sensor/kubernetes/listener/resources/selector.go

```diff
@@ -4,42 +4,107 @@ import (
 type selector interface {
-	Matches(labels.Labels) bool
+	Matches(labelsWithLen) bool
+}
+
+type labelWithLenImpl struct {
+	labels map[string]string
+}
+
+func (l labelWithLenImpl) Has(label string) bool {
+	_, exists := l.labels[label]
+	return exists
+}
+
+func (l labelWithLenImpl) Get(label string) string {
+	return l.labels[label]
+}
+
+func (l labelWithLenImpl) Len() uint {
+	return uint(len(l.labels))
+}
+
+func createLabelsWithLen(labels map[string]string) labelWithLenImpl {
+	return labelWithLenImpl{labels}
+}
+
+// SelectorWrap holds a selector and information allowing for additional checks before matching
+type selectorWrap struct {
+	selector  labels.Selector
+	numLabels uint
+	matchNil  bool
+}
+
+func (s selectorWrap) Matches(labels labelsWithLen) bool {
+	if s.numLabels > labels.Len() {
+		return false
+	}
+	if s.numLabels == 0 {
+		return s.matchNil
+	}
+	return s.selector.Matches(labels)
 }
 
 // selectorDisjunction is the disjunction (logical or) of a list of selectors.
 type selectorDisjunction []selector
 
-func (d selectorDisjunction) Matches(labels2 labels.Labels) bool {
+func (d selectorDisjunction) Matches(labels labelsWithLen) bool {
 	for _, sel := range d {
-		if sel.Matches(labels2) {
+		if sel.Matches(labels) {
 			return true
 		}
 	}
 	return false
 }
 
-// or returns the logical or of the given selectors.
+// or returns the logical or of the given SelectorWrappers.
 func or(sels ...selector) selector {
 	return selectorDisjunction(sels)
 }
 
-// SelectorFromMap converts the given map to a selector. It correctly translates a `nil` map to a `nothing` collector,
-// as is, e.g., used for headless services.
-func SelectorFromMap(labelsMap map[string]string) labels.Selector {
-	if len(labelsMap) == 0 {
-		return labels.Nothing()
+type selectorWrapOption func(*selectorWrap)
+
+// emptyMatchesNothing means that a set with no labels should not match with anything
+func emptyMatchesNothing() selectorWrapOption {
+	return func(sw *selectorWrap) {
+		sw.matchNil = false
+	}
+}
+
+// emptyMatchesEverything means that a set with no labels should match with everything
+func emptyMatchesEverything() selectorWrapOption {
+	return func(sw *selectorWrap) {
+		sw.matchNil = true
 	}
-	return labels.SelectorFromSet(labels.Set(labelsMap))
 }
 
-// MatcherOrEverything converts the given map to a selector. If the map is `nil` or empty it will translate to `everything`
-// selector. This is needed in cases like Network Policies where an empty PodSelector matches everything in the namespace.
-func MatcherOrEverything(labelsMap map[string]string) labels.Selector {
-	if len(labelsMap) == 0 {
-		return labels.Everything()
+// CreateSelector returns a SelectorWrapper for the given map of labels; matchNil determines whether
+// an empty set of labels matches everything or nothing.
+func createSelector(labelsMap map[string]string, opts ...selectorWrapOption) selectorWrap {
+	var selWrapper selectorWrap
+	selWrapper.matchNil = false
```

Comment:
This is optional - all uninitialized boolean fields default to false.

Optionally, you could do:

```go
selWrapper := selectorWrap{matchNil: false}
```

Reply:
If that is your preference, then I would suggest using `selWrapper := selectorWrap{matchNil: false}`

Reply:
Okay